### PR TITLE
Create the Social media repository

### PR DESCRIPTION
**Pull request by GabrielFleischer**
This is the first part of the decoupling of the social media feature from the Lao object.

This PR only create the repository, it is not used anywhere in the codebase yet.

While writing tests, I had to change the chirp object to an immutable one. So it is part of the PR 😅

### CODE REVIEW

File: fe2-android/app/src/test/unit/java/com/github/dedis/popstellar/repository/SocialMediaRepositoryTest.java [hunk 1]
```diff
@@ -0,0 +1,128 @@
...
+          "This is another chirp !",
+          1003,
+          new MessageID(""));
+
+  @Test
+  public void addingAChirpAfterSubscriptionUpdatesIds() {
+    SocialMediaRepository repo = new SocialMediaRepository();
+    TestObserver<Set<MessageID>> ids = repo.getChirpsOfLao(LAO_ID).test();
+    // make sure the current last element is an empty set
+    ids.assertValueAt(ids.valueCount() - 1, emptySet());
```
Comment (pierluca) [inline fe2-android/app/src/test/unit/java/com/github/dedis/popstellar/repository/SocialMediaRepositoryTest.java:40]: do we really need `valueCount() -1`  , can't we just say `0`, `1`, `2`, etc. ? It'd be A LOT more readable.

File: fe2-android/app/src/test/unit/java/com/github/dedis/popstellar/repository/SocialMediaRepositoryTest.java [hunk 2]
```diff
@@ -0,0 +1,128 @@
...
+    ids.assertValueAt(ids.valueCount() - 1, singleton(CHIRP_1.getId()));
+  }
+
+  @Test
+  public void addingChirpBeforeSubscriptionUpdateIds() {
+    SocialMediaRepository repo = new SocialMediaRepository();
+    repo.addChirp(LAO_ID, CHIRP_1);
+    TestObserver<Set<MessageID>> ids = repo.getChirpsOfLao(LAO_ID).test();
+    // The value at subscription contains the first chirp's id
+    ids.assertValueAt(ids.valueCount() - 1, singleton(CHIRP_1.getId()));
```
Comment (pierluca) [inline fe2-android/app/src/test/unit/java/com/github/dedis/popstellar/repository/SocialMediaRepositoryTest.java:52]: what does `singleton()` do?  do we need it ?

File: fe2-android/app/src/test/unit/java/com/github/dedis/popstellar/repository/SocialMediaRepositoryTest.java [hunk 3]
```diff
@@ -0,0 +1,128 @@
+package com.github.dedis.popstellar.repository;
```
Comment (pierluca) [inline fe2-android/app/src/test/unit/java/com/github/dedis/popstellar/repository/SocialMediaRepositoryTest.java:1]: these tests are overalls not readable 😢 they're not bad at all, just not very readable.
I'll try to provide some comments to help make it cleaner 😄

File: fe2-android/app/src/test/unit/java/com/github/dedis/popstellar/repository/SocialMediaRepositoryTest.java [hunk 4]
```diff
@@ -0,0 +1,128 @@
...
+    Chirp invalidChirp =
+        new Chirp(
+            CHIRP_1.getId(),
+            generatePublicKey(),
+            "This is another chirp !",
+            1003,
+            new MessageID(""));
+    repo.addChirp(LAO_ID, invalidChirp);
+    chirp.assertValueAt(chirp.valueCount() - 1, CHIRP_1);
+  }
```
Comment (pierluca) [inline fe2-android/app/src/test/unit/java/com/github/dedis/popstellar/repository/SocialMediaRepositoryTest.java:83]: here and elsewhere, but even more so for this test, it would help here to clearly separate the `arrange` (pre-condition asserts included) / `act` / `assert` sections of the test
```suggestion
  public void addChirpWithExistingIdHasNoEffect() throws UnknownChirpException {
    // arrange
    SocialMediaRepository repo = new SocialMediaRepository();
    repo.addChirp(LAO_ID, CHIRP_1);
    
    TestObserver<Chirp> chirp = repo.getChirp(LAO_ID, CHIRP_1.getId()).test();
    chirp.assertValueAt(chirp.valueCount() - 1, CHIRP_1);
    
    // act
    Chirp invalidChirp =
        new Chirp(
            CHIRP_1.getId(),
            generatePublicKey(),
            "This is another chirp !",
            1003,
            new MessageID(""));
    repo.addChirp(LAO_ID, invalidChirp);
    
    // assert
    chirp.assertValueAt(chirp.valueCount() - 1, CHIRP_1);
  }
```

File: fe2-android/app/src/main/java/com/github/dedis/popstellar/model/objects/Chirp.java [hunk 5]
```diff
@@ -6,32 +6,57 @@
...
   private static final int MAX_CHIRP_CHARS = 300;
 
-  private MessageID id;
-  private Channel channel;
+  private final MessageID id;
+  private final PublicKey sender;
+  private final String text;
+  private final long timestamp;
+  private final boolean isDeleted;
+  private final MessageID parentId;
```
Comment (pierluca) [inline fe2-android/app/src/main/java/com/github/dedis/popstellar/model/objects/Chirp.java:21]: what's the parent ID ? I don't recall this in the specification.

File: fe2-android/app/src/test/unit/java/com/github/dedis/popstellar/repository/SocialMediaRepositoryTest.java (hunk 4 above)
Comment (pierluca) [reply fe2-android/app/src/test/unit/java/com/github/dedis/popstellar/repository/SocialMediaRepositoryTest.java:83]: Naive/rhetorical question: is this the case because the chirp is invalid or because we ignore a new ID ? :-)

It might be helpful to document in the repository somewhere that we expect only valid data to be provided at this level and that data validation has already happened before.

**Review by pierluca — CHANGES_REQUESTED** [review]
Overalls, this looks pretty good. However, the tests are hard to follow.

File: fe2-android/app/src/test/unit/java/com/github/dedis/popstellar/repository/SocialMediaRepositoryTest.java (hunk 1 above)
Comment (GabrielFleischer) [reply fe2-android/app/src/test/unit/java/com/github/dedis/popstellar/repository/SocialMediaRepositoryTest.java:40]: To me, this is impmentation defined, I didn't want to force the underlying implementation to have nothing published before and exactly one published after.

I would have love a `assertLastValue()` and I am thinking about creating it, but it will look like `assertLastValueIs(observer, value)`

File: fe2-android/app/src/test/unit/java/com/github/dedis/popstellar/repository/SocialMediaRepositoryTest.java (hunk 2 above)
Comment (GabrielFleischer) [reply fe2-android/app/src/test/unit/java/com/github/dedis/popstellar/repository/SocialMediaRepositoryTest.java:52]: Not that much, it is a usual way to create a set containing one element.

For readability, I could use the function `setOf()` yes

File: fe2-android/app/src/main/java/com/github/dedis/popstellar/model/objects/Chirp.java (hunk 5 above)
Comment (GabrielFleischer) [reply fe2-android/app/src/main/java/com/github/dedis/popstellar/model/objects/Chirp.java:21]: I don't know for sure, I only followed the already existing impmentation.

My guess is that it is the ID for the Chirp it is "replying" to.

File: fe2-android/app/src/test/unit/java/com/github/dedis/popstellar/repository/SocialMediaRepositoryTest.java (hunk 3 above)
Comment (GabrielFleischer) [reply fe2-android/app/src/test/unit/java/com/github/dedis/popstellar/repository/SocialMediaRepositoryTest.java:1]: Yes, understandable 😅

I had a hard time writing the reactive tests, I almost never did it before.

File: fe2-android/app/src/test/unit/java/com/github/dedis/popstellar/repository/SocialMediaRepositoryTest.java (hunk 4 above)
Comment (GabrielFleischer) [reply fe2-android/app/src/test/unit/java/com/github/dedis/popstellar/repository/SocialMediaRepositoryTest.java:83]: To me, it can occur if a Chirp message is doubled.

And yes it is expected that the ID is valide at this point.

I will add it is the documentation.

**Review by pierluca — APPROVED** [review]
So much better. Good job !
